### PR TITLE
Gh integ

### PR DESCRIPTION
**Pull request by greathorn-pzl**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
Initial version of the integration with the GreatHorn product

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Demisto
- [ ] 5.0.0
- [ ] 5.5.0
- [x] 6.0.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: Packs/GreatHorn/Integration/GreatHorn/GreatHorn.yml [hunk 1]
```diff
@@ -0,0 +1,868 @@
...
+      description: Whether or not to include full message headers in the War Room output.
+      defaultValue: "false"
+    - name: showalllinks
+      auto: PREDEFINED
+      predefined:
+      - "true"
+      - "false"
+      description: Whether or not to show all links in the War Room output. When false only suspicious and malicious links will be returned to the War Room.
+      defaultValue: "false"
+    outputs:
```
Comment (abaumgarten) [inline Packs/GreatHorn/Integration/GreatHorn/GreatHorn.yml:59]: Please add `description` to all of your outputs (in all the commands)

File: Packs/GreatHorn/Integration/GreatHorn/GreatHorn.yml [hunk 2]
```diff
@@ -0,0 +1,868 @@
...
+      type: Unknown
+    - contextPath: GreatHorn.Message.headers.content-type
+      description: ""
+      type: Unknown
+  - name: gh-search-message
+    description: Search for message based on filtering input
+    arguments:
+    - name: fields
+      description: The fields to include in the response. By default, all fields are
+        returned.
```
Comment (abaumgarten) [inline Packs/GreatHorn/Integration/GreatHorn/GreatHorn.yml:395]: Please note: If a command uses argToList function the argument `isArray` should be set to true.

File: Packs/GreatHorn/Integration/GreatHorn/GreatHorn.yml [hunk 3]
```diff
@@ -0,0 +1,868 @@
...
+      description: ""
+      type: String
+    - contextPath: GreatHorn.Remediation.success
+      description: ""
+      type: Number
+  - name: gh-get-policy
+    description: Retrieve details about the policy specified
+    arguments:
+    - name: policyid
+      description: The GreatHorn policy ID
```
Comment (abaumgarten) [inline Packs/GreatHorn/Integration/GreatHorn/GreatHorn.yml:831]: Please note: If a command uses argToList function the argument `isArray` should be set to true.

File: Packs/GreatHorn/Integration/GreatHorn/GreatHorn.yml [hunk 4]
```diff
@@ -0,0 +1,868 @@
...
+script:
+  script: ''
+  type: python
+  commands:
+  - name: gh-get-message
+    description: Return message details for the specified event
+    arguments:
+    - name: id
+      required: true
+      description: GreatHorn eventId, multiple values supported via CSV
```
Comment (abaumgarten) [inline Packs/GreatHorn/Integration/GreatHorn/GreatHorn.yml:43]: Please note: If a command uses argToList function the argument `isArray` should be set to true.

File: Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py [hunk 5]
```diff
@@ -0,0 +1,540 @@
...
+
+def main() -> None:
+    """main function, parses params and runs command functions
+
+    :return:
+    :rtype:
+    """
+
+    api_key = demisto.params().get('apikey')
+    fetch_type = demisto.params().get('fetch_type')
```
[the remark this answers is not in the document]
Comment (abaumgarten) [reply Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py:462]: In addition, `first_fetch` param should be inside the integration configuration.

**Review by abaumgarten — COMMENTED** [review]
@greathorn-pzl, Thank you for your contribution!
Please see my comments below, I am looking forward to talking to you today.
Please feel free to dm me on slack if you need any help or if something is not clear.
Some general comments:
1. Please make sure that email addresses that are part of your test_data files are fake. If those addresses are not fake, make sure that you understand they can be seen as our repo is public.
2. Please provide README files for your integration and for your playbook. You can use demisto-sdk generate-docs in order to do that automatically.
Thank you !

**Comment by abaumgarten** [comment]
Hi, one more thing please, because you are a Partner, you should add your pack a publisher logo. This logo will be displayed on the pack details view (in MarketPlace). 
In order to add the logo, you should edit the `pack_metadata.json` file: add a link to a publisher logo under a key called `authorImage`.
More information about the logo you can find here: https://xsoar.pan.dev/docs/integrations/packs-format#author_imagepng.
Thank you.

File: Packs/GreatHorn/Integration/GreatHorn/GreatHorn.yml [hunk 6]
```diff
@@ -0,0 +1,868 @@
+commonfields:
+  id: GreatHorn
+  version: -1
+fromversion: 5.0.0
```
[the remark this answers is not in the document]
Comment (greathorn-pzl) [reply Packs/GreatHorn/Integration/GreatHorn/GreatHorn.yml]: Updated.

File: Packs/GreatHorn/Integration/GreatHorn/GreatHorn.yml [hunk 7]
```diff
@@ -0,0 +1,868 @@
...
+      - asc
+      description: Indicates if the sort direction is ascending or descending. Default
+        is descending.
+      defaultValue: desc
+    outputs:
+    - contextPath: GreatHorn.Message.origin
+      description: ""
+      type: String
+    - contextPath: GreatHorn.Message.status
+      description: ""
```
[the remark this answers is not in the document]
Comment (greathorn-pzl) [reply Packs/GreatHorn/Integration/GreatHorn/GreatHorn.yml]: There context paths are in separate commands so I would think they would separate scopes.  The fact is that we return the same type of data from two commands so the contexts would be similar.

File: Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py [hunk 8]
```diff
@@ -0,0 +1,540 @@
...
+        incident['rawJSON'] = json.dumps(event)
+        client.remediate_message("review", {"eventId": event.get("eventId", 0)})
+        incidents.append(incident)
+    return incidents
+
+
+''' MAIN FUNCTION '''
+
+
+def main() -> None:
```
[the remark this answers is not in the document]
Comment (greathorn-pzl) [reply Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py]: done

File: Packs/GreatHorn/pack_metadata.json [hunk 9]
```diff
@@ -0,0 +1,17 @@
+
+{
+  "name": "GreatHorn",
+  "description": "The only cloud-native security platform that stops targeted social engineering and phishing attacks on cloud email platforms like Office 365 and G Suite.",
+  "support": "xsoar",
```
[the remark this answers is not in the document]
Comment (greathorn-pzl) [reply Packs/GreatHorn/pack_metadata.json]: Needed a lower case P, but done.

File: Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py [hunk 10]
```diff
@@ -0,0 +1,540 @@
...
+
+    :return: 'ok' if test passed, anything else will fail the test.
+    :rtype: ``str``
+    """
+
+    try:
+        client.get_policy()
+    except DemistoException as e:
+        if 'Forbidden' in str(e):
+            return 'Authorization Error: make sure API Key is correctly set'
```
[the remark this answers is not in the document]
Comment (greathorn-pzl) [reply Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py]: done

File: Packs/GreatHorn/IncidentTypes/incidenttype-GreatHorn_Phishing.json [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+    "locked": false,
+    "name": "GreatHorn Phishing",
+    "playbookId": "GreatHornGetEventStatus",
+    "readonly": false,
+    "reputationCalc": 0,
+    "system": false,
+    "version": -1,
+    "weeks": 0,
+    "weeksR": 0,
+    "fromVersion": "5.0.0"
```
[the remark this answers is not in the document]
Comment (greathorn-pzl) [reply Packs/GreatHorn/IncidentTypes/incidenttype-GreatHorn_Phishing.json]: Done

File: Packs/GreatHorn/Integration/GreatHorn/GreatHorn.yml (hunk 1 above)
Comment (greathorn-pzl) [reply Packs/GreatHorn/Integration/GreatHorn/GreatHorn.yml:60]: Done.  I have made a change to the email headers so we don't specify every one.  That is one thing I want to run through.

File: Packs/GreatHorn/Integration/GreatHorn/GreatHorn.yml (hunk 2 above)
Comment (greathorn-pzl) [reply Packs/GreatHorn/Integration/GreatHorn/GreatHorn.yml]: done

File: Packs/GreatHorn/Integration/GreatHorn/GreatHorn.yml (hunk 3 above)
Comment (greathorn-pzl) [reply Packs/GreatHorn/Integration/GreatHorn/GreatHorn.yml]: done

File: Packs/GreatHorn/Integration/GreatHorn/GreatHorn.yml (hunk 4 above)
Comment (greathorn-pzl) [reply Packs/GreatHorn/Integration/GreatHorn/GreatHorn.yml]: done

**Comment by greathorn-pzl** [comment]
> Hi, one more thing please, because you are a Partner, you should add your pack a publisher logo. This logo will be displayed on the pack details view (in MarketPlace).
> In order to add the logo, you should edit the `pack_metadata.json` file: add a link to a publisher logo under a key called `authorImage`.
> More information about the logo you can find here: https://xsoar.pan.dev/docs/integrations/packs-format#author_imagepng.
> Thank you.

We have a file called GreatHorn_image.png  in the integration directory per the instructions and that was sufficient for it to display when I loaded the integration before.  Do I need to add it to the pack_metadata.json file too?

**Comment by abaumgarten** [comment]
> > Hi, one more thing please, because you are a Partner, you should add your pack a publisher logo. This logo will be displayed on the pack details view (in MarketPlace).
> > In order to add the logo, you should edit the `pack_metadata.json` file: add a link to a publisher logo under a key called `authorImage`.
> > More information about the logo you can find here: https://xsoar.pan.dev/docs/integrations/packs-format#author_imagepng.
> > Thank you.
> 
> We have a file called GreatHorn_image.png in the integration directory per the instructions and that was sufficient for it to display when I loaded the integration before. Do I need to add it to the pack_metadata.json file too?

So I think that it should be all right and do don't need to change pack_metadata.json. @mchasepan Am I right?

File: Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py [hunk 12]
```diff
@@ -0,0 +1,540 @@
...
+
+
+def gh_get_message_command(client: Client, args: Dict[str, Any]) -> CommandResults:
+    ghid = argToList(args.get('id'))
+    includeheaders = args.get('includeheaders', "false").lower() == "true"
+    showalllinks = args.get('showalllinks', "false").lower() == "true"
+    result = client.search_events({"filters": [{"eventId": ghid}]})
+
+    if len(result.get("results", [])) > 0:
+        message = result.get("results", [None])[0]
```
[the remark this answers is not in the document]
Comment (greathorn-pzl) [reply Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py:302]: Making the suggested change resulted in a error so I did not make the change.

File: Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py [hunk 13]
```diff
@@ -0,0 +1,540 @@
...
+        else:
+            headers_md = ""
+
+        message_md = envelope_md + authentication_md + scores_md + links_md + files_md + policies_md + headers_md
+
+        return CommandResults(
+            readable_output=message_md,
+            outputs_prefix='GreatHorn.Message',
+            outputs_key_field='eventId',
+            outputs=result.get("results", [None])[0]
```
[the remark this answers is not in the document]
Comment (greathorn-pzl) [reply Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py:405]: Same comment as above.

File: Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py [hunk 14]
```diff
@@ -0,0 +1,540 @@
...
+
+        elif demisto.command() == 'gh-get-message':
+            return_results(gh_get_message_command(client, demisto.args()))
+
+        elif demisto.command() == 'gh-get-policy':
+            return_results(gh_get_policy_command(client, demisto.args()))
+
+        elif demisto.command() == 'gh-set-policy':
+            return_results(gh_set_policy_command(client, demisto.args()))
+
```
[the remark this answers is not in the document]
Comment (greathorn-pzl) [reply Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py:500]: Thanks for catching this.  Adding it in.

[the remark this answers is not in the document]
Comment (abaumgarten) [reply Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py:500]: I do not see this in the updated yml file. Please make sure you pushed the commit.

[the remark this answers is not in the document]
Comment (abaumgarten) [reply Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py:500]: I do not see this in the updated yml file. Please make sure you pushed the commit.

File: Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py [hunk 15]
```diff
@@ -0,0 +1,540 @@
...
+    return CommandResults(
+        readable_output=policies_md,
+        outputs_prefix='GreatHorn.Policy',
+        outputs_key_field='id',
+        outputs=results
+    )
+
+
+def gh_get_message_command(client: Client, args: Dict[str, Any]) -> CommandResults:
```
[the remark this answers is not in the document]
Comment (greathorn-pzl) [reply Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py]: We will look at that for the next version.

File: Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (greathorn-pzl) [reply Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py]: The changes are there for review now.

File: Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py [hunk 16]
```diff
@@ -0,0 +1,540 @@
...
+                                      "quarantinerelease": [],
+                                      "quarantinedeny": [],
+                                      "removeattachments": [],
+                                      "review": []
+                                      }
+    if args.get('action', "").lower() not in valid_actions.keys():
+        raise ValueError('Invalid action "{}" specified'.format(args.get('action')))
+    action = args.get('action', "").lower()
+    for arg in valid_actions.get(action, []):
+        if args.get(arg) is None:
```
[the remark this answers is not in the document]
Comment (greathorn-pzl) [reply Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py]: I think that the check is not needed, but I don't think it is a bug.   As you note, there are no arguments to the actions in this case that need to be found, so we will never get to line 175.   If this changes in the future we would have the logic setup to handle the case in question.

File: Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py [hunk 17]
```diff
@@ -0,0 +1,540 @@
...
+    policy_id = args.get('policiyid', "")
+    policy_json = json.loads(args.get('policyjson', {}))
+    if update_method not in ['patch', 'put']:
+        raise ValueError("Invalid updatemethod specified, please use either put or patch.")
+    results = client.set_policy(policy_id, update_method, policy_json)
+    if results.get("success") is True:
+        human_readable = "Update applied successfully to policy {}".format(args.get("policiyid"))
+    results['id'] = args.get('policiyid')
+    return CommandResults(
+        readable_output=human_readable,
```
[the remark this answers is not in the document]
Comment (greathorn-pzl) [reply Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py]: done

File: Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py [hunk 18]
```diff
@@ -0,0 +1,540 @@
...
+    if results.get("success") is True:
+        human_readable = "Revert action {} applied successfully to message {}".format(action, args.get("eventId"))
+    else:
+        if results.get("reason") == "alreadyDone":
+            human_readable = "Revert action {} has already been performed on message {}".format(action, args.get("eventId"))
+    results['action'] = action
+    results['eventId'] = args.get('eventId')
+
+    return CommandResults(
+        readable_output=human_readable,
```
[the remark this answers is not in the document]
Comment (greathorn-pzl) [reply Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py]: done

File: Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py [hunk 19]
```diff
@@ -0,0 +1,540 @@
...
+    return 'ok'
+
+
+def gh_search_message_command(client: Client, args: Dict[str, Any]) -> CommandResults:
+    fields = argToList(args.get('fields'))
+    limit = args.get('limit')
+    sort = args.get('sort')
+    sortDir = args.get('sortDir')
+    offset = args.get('offset')
+    filters = json.loads(args.get('filters', "[]"))
```
[the remark this answers is not in the document]
Comment (greathorn-pzl) [reply Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py]: I updated the description in the yml file but had some trouble with the checking of the argument.

File: Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py [hunk 20]
```diff
@@ -0,0 +1,540 @@
...
+                                                       "Subject", "Policy Hits", "Policy Actions",
+                                                       "Occurred", "Has Attachments", "Has Links"])
+    else:
+        events_md = tableToMarkdown("Events", results.get("results", []), fields)
+
+    result = {'Message': results.get("results"), 'SearchCount': results.get("total")}
+
+    return CommandResults(
+        readable_output=events_md,
+        outputs_prefix='GreatHorn',
```
[the remark this answers is not in the document]
Comment (greathorn-pzl) [reply Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py]: done

File: Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (greathorn-pzl) [reply Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py]: Don't know what you mean by first_fetch.   I didn't see it in the documentation either.

File: Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py [hunk 21]
```diff
@@ -0,0 +1,540 @@
...
+        elif demisto.command() == 'gh-remediate-message':
+            return_results(gh_remediate_message_command(client, demisto.args()))
+
+        elif demisto.command() == 'gh-revert-remediate-message':
+            return_results(gh_revert_remediate_message_command(client, demisto.args()))
+
+        elif demisto.command() == 'gh-search-message':
+            return_results(gh_search_message_command(client, demisto.args()))
+
+        elif demisto.command() == 'fetch-incidents':
```
[the remark this answers is not in the document]
Comment (greathorn-pzl) [reply Packs/GreatHorn/Integration/GreatHorn/GreatHorn.py]: Added.

File: Packs/GreatHorn/Integrations/GreatHorn/GreatHorn.py [hunk 22]
```diff
@@ -0,0 +1,546 @@
...
+        return self._http_request(
+            method=update_method,
+            url_suffix='/policy/{}'.format(policy_id),
+            json_data=policy_json)
+
+    def search_events(self, args: Optional[dict]) -> Dict[str, Any]:
+        return self._http_request(
+            method='POST',
+            url_suffix='/search/events',
+            json_data=args)
```
Comment (abaumgarten) [inline Packs/GreatHorn/Integrations/GreatHorn/GreatHorn.py:45]: Just to make sure, when do you use the argument `max_fetch`? I might be wrong but it seems like you do not use it to limit the number of incidents that you have got from the API.

**Review by abaumgarten — COMMENTED** [review]
@greathorn-pzl, thank you for your fixes,  please see my new few comments bellow. Please let me know if you need my help.

File: Packs/GreatHorn/Integrations/GreatHorn/GreatHorn.py (hunk 22 above)
Comment (greathorn-pzl) [reply Packs/GreatHorn/Integrations/GreatHorn/GreatHorn.py:45]: max_fetch is passed as limit to  gh_get_phish_reports_command and gh_get_quarantine_release_command which use it to set the argument limit which is passed to search_events.

Comment (abaumgarten) [reply Packs/GreatHorn/Integrations/GreatHorn/GreatHorn.py:45]: What will happen with those incidents that are beyond the limit? For example, if the limit is 100 and there were 102 incidents? Do you save the last 2 incidents somewhere? Do you want to let the user know that there were incidents that are not being displayed?

File: Packs/GreatHorn/Integrations/GreatHorn/GreatHorn.py [hunk 23]
```diff
@@ -0,0 +1,546 @@
...
+                    elif fetch_type == "quarantine":
+                        incidents = gh_get_quarantine_release_command(client, max_fetch)
+                    else:
+                        incidents_phish = gh_get_phish_reports_command(client, max_fetch)
+                        incidents_quarantine = gh_get_quarantine_release_command(client, max_fetch)
+                        incidents = incidents_phish
+                        incidents.extend(incidents_quarantine)
+                    demisto.incidents(incidents)
+                counter += 1
+                demisto.setLastRun({'max_phish_id': str(counter)})
```
[the remark this answers is not in the document]
Comment (greathorn-pzl) [reply Packs/GreatHorn/Integrations/GreatHorn/GreatHorn.py:536]: The issue with setting last run based on time is that we are not using a time based fetch.   Instead the code is written to fetch incidents every third time it is run.  I realize that this is not how the documentation recommends using LastRun, but I do see other occurrences of fetchs being written without using a date for LastRun.

File: Tests/conf.json [hunk 24]
```diff
@@ -3004,6 +3004,10 @@
             "integrations": "IBM Resilient Systems",
             "playbookID": "IBM Resilient Systems Test"
         },
+	{
+            "integrations": "GreatHorn",
+            "playbookID": "TestGreatHornPlaybook"
+        },
```
[the remark this answers is not in the document]
Comment (greathorn-pzl) [reply Tests/conf.json]: done

**Comment by abaumgarten** [comment]
Status: I have asked @greathorn-pzl some questions about `fetch incidents` flow. I am waiting for a response if those changes (i.e `fetch incidents` flow) will be part of this version or in the next one instead. There are no more changes besides `fetch incidents` ones.
@guykeller - FYI

**Comment by abaumgarten** [comment]
> Status: I have asked @greathorn-pzl some questions about `fetch incidents` flow. I am waiting for a response if those changes (i.e `fetch incidents` flow) will be part of this version or in the next one instead. There are no more changes besides `fetch incidents` ones.
> @guykeller - FYI

Update - a call about fetch incidents is scheduled for tomorrow .

**Comment by abaumgarten** [comment]
updated status: After a call with @greathorn-pzl and her colleges, @guykeller, and me, we have decided that the first version will not include `fetch incidents`.

**Review by abaumgarten — COMMENTED** [review]
Please see my only comment below

File: Packs/GreatHorn/pack_metadata.json [hunk 25]
```diff
@@ -0,0 +1,17 @@
+
+{
+  "name": "GreatHorn",
+  "description": "The only cloud-native security platform that stops targeted social engineering and phishing attacks on cloud email platforms like Office 365 and G Suite.",
+  "support": "partner",
+  "currentVersion": "1.0.0",
+  "author": "GreatHorn",
+  "url": "",
```
[the remark this answers is not in the document]
Comment (greathorn-pzl) [reply Packs/GreatHorn/pack_metadata.json:8]: do you mean for our product support or for our corporate site?

[the remark this answers is not in the document]
Comment (abaumgarten) [reply Packs/GreatHorn/pack_metadata.json:8]: should be a link to the vendor site

**Review by abaumgarten — APPROVED** [review]
@greathorn-pzl Approved. Thank you for your contribution, great job!

**Pull request merged**